File: bigo/quick_sort.c
#include <stdio.h>

// 平均 nlogn  最差 n^2
int quick_sort(int tmp[], int left, int right)

{

    int i,j,key,t;
    if(left > right)
        return;

    i = left;
    j = right;
    key = tmp[left];
    while(i != j){ 
        while(tmp[j]>=key && i < j ){
            j--;
        }   
        while(tmp[i] <= key && i < j){ 
            i++;
        }   
        if (i < j){ 
            t = tmp[i];
            tmp[i] = tmp[j];
            tmp[j] = t;
        }   

    }   
    tmp[left] = tmp[i];
    tmp[i] = key;

    quick_sort(tmp,left,i-1);
    quick_sort(tmp,i+1,right);



}

int main()
{
    int tmp[] = {10,2,4,5,7,1,9,8,6,0,3};
    int len = sizeof(tmp)/sizeof(int);
    printf("len=%d\n",len);
    quick_sort(tmp,0,len-1);
    int i=0;
    for(; i<len; i++){
        printf("%d\n", tmp[i]);
    }

    return 0;
}

File: bigo/quick_sort.java
public class quickSort {
    static void swap(int [] src, int i, int j){
        int t = src[i];
        src[i]=src[j];
        src[j]=t;
    }
    static void  quicksort(int [] src, int left, int right){
        int i=left;
        int j=right;

        if(left >= right){
            return;
        }

        int key = src[left];
        while(i!=j){
            while(src[j] >= key && i<j){
                j--;
            }
            while(src[i] <= key && i <j)
                i++;
            if(i<j){
                swap(src,i,j);
            }
        }
        src[left] = src[i];
        src[i]= key;

        quicksort(src,left,i-1);
        quicksort(src,i+1,right);

    }

    public static void main(String[] args) {
        int [] src = {9,8,7,6,5,4,3,3,2,2,1,1,0,-1};
        quicksort(src,0,src.length-1);
        for (int t: src
             ) {
            System.out.println(t);
        }

    }
}
